File: src/lexer.ts
// Clank lexer — tokenizes source into structured tokens
// ASCII-only, # comments, no semicolons

import type { Loc } from "./ast.js";

// ── Token types ──

export type TokenTag =
  | "int" | "rat" | "str" | "bool"
  | "ident" | "keyword"
  | "op" | "delim"
  | "eof";

export type Token = { tag: TokenTag; value: string; loc: Loc };

export type LexError = {
  code: string;
  message: string;
  location: Loc;
  context: string;
};

// ── Keywords ──

const KEYWORDS = new Set([
  "let", "in", "for", "fn", "if", "then", "else", "match", "do", "type", "effect",
  "affine", "handle", "resume", "perform", "mod", "use", "pub", "clone", "true", "false",
  "interface", "impl", "Self", "deriving", "where", "opaque", "return", "test", "alias",
]);

// ── Multi-char operators (longest-match order) ──

const MULTI_OPS = [
  "==", "!=", "<=", ">=", "&&", "||", "++", "|>", "=>", "<-", "->",
];

const SINGLE_OPS = new Set([
  "+", "-", "*", "/", "%", "<", ">", "!", "=", "\\",
]);

const DELIMITERS = new Set([
  "{", "}", "(", ")", "[", "]", ",", ":", ".", "|", "&",
]);

// ── Helpers ──

function isAlpha(ch: string): boolean {
  return (ch >= "a" && ch <= "z") || (ch >= "A" && ch <= "Z");
}

function isDigit(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}

function isIdentChar(ch: string): boolean {
  return isAlpha(ch) || isDigit(ch) || ch === "_" || ch === "-";
}

function isWhitespace(ch: string): boolean {
  return ch === " " || ch === "\t" || ch === "\n" || ch === "\r";
}

// ── Lexer ──

export function lex(source: string): Token[] | LexError {
  const tokens: Token[] = [];
  let pos = 0;
  let line = 1;
  let col = 1;

  function loc(): Loc {
    return { line, col };
  }

  function endLoc(): { end_line: number; end_col: number } {
    return { end_line: line, end_col: col };
  }

  function peek(): string {
    return pos < source.length ? source[pos] : "";
  }

  function advance(): string {
    const ch = source[pos++];
    if (ch === "\n") { line++; col = 1; } else { col++; }
    return ch;
  }

  function error(msg: string, l?: Loc): LexError {
    const at = l ?? loc();
    return { code: "E001", message: msg, location: at, context: source.slice(Math.max(0, pos - 20), pos + 20) };
  }

  while (pos < source.length) {
    const ch = peek();

    // Whitespace
    if (isWhitespace(ch)) { advance(); continue; }

    // Comment
    if (ch === "#") {
      while (pos < source.length && peek() !== "\n") advance();
      continue;
    }

    const start = loc();

    // String literal
    if (ch === '"') {
      advance(); // opening quote
      let value = "";
      while (pos < source.length && peek() !== '"') {
        if (peek() === "\\") {
          advance();
          const esc = advance();
          if (esc === "n") value += "\n";
          else if (esc === "t") value += "\t";
          else if (esc === "\\") value += "\\";
          else if (esc === '"') value += '"';
          else return error(`invalid escape \\${esc}`, start);
        } else {
          value += advance();
        }
      }
      if (pos >= source.length) return error("unterminated string", start);
      advance(); // closing quote
      tokens.push({ tag: "str", value, loc: { ...start, ...endLoc() } });
      continue;
    }

    // Number literal (int or rat)
    if (isDigit(ch)) {
      let num = "";
      while (pos < source.length && isDigit(peek())) num += advance();
      if (peek() === "." && pos + 1 < source.length && isDigit(source[pos + 1])) {
        num += advance(); // the dot
        while (pos < source.length && isDigit(peek())) num += advance();
        tokens.push({ tag: "rat", value: num, loc: { ...start, ...endLoc() } });
      } else {
        tokens.push({ tag: "int", value: num, loc: { ...start, ...endLoc() } });
      }
      continue;
    }

    // Identifier or keyword
    if (isAlpha(ch) || ch === "_") {
      let word = "";
      while (pos < source.length && isIdentChar(peek())) word += advance();
      const end = endLoc();
      if (word === "true" || word === "false") {
        tokens.push({ tag: "bool", value: word, loc: { ...start, ...end } });
      } else if (KEYWORDS.has(word)) {
        tokens.push({ tag: "keyword", value: word, loc: { ...start, ...end } });
      } else {
        tokens.push({ tag: "ident", value: word, loc: { ...start, ...end } });
      }
      continue;
    }

    // Multi-char operators (try longest match first)
    if (pos + 1 < source.length) {
      const two = source.slice(pos, pos + 2);
      if (MULTI_OPS.includes(two)) {
        advance(); advance();
        tokens.push({ tag: "op", value: two, loc: { ...start, ...endLoc() } });
        continue;
      }
    }

    // Range operators: ..= (3-char) and .. (2-char)
    if (ch === "." && pos + 1 < source.length && source[pos + 1] === ".") {
      if (pos + 2 < source.length && source[pos + 2] === "=") {
        advance(); advance(); advance();
        tokens.push({ tag: "op", value: "..=", loc: { ...start, ...endLoc() } });
      } else {
        advance(); advance();
        tokens.push({ tag: "op", value: "..", loc: { ...start, ...endLoc() } });
      }
      continue;
    }

    // Single-char operators
    if (SINGLE_OPS.has(ch)) {
      advance();
      tokens.push({ tag: "op", value: ch, loc: { ...start, ...endLoc() } });
      continue;
    }

    // Delimiters
    if (DELIMITERS.has(ch)) {
      advance();
      tokens.push({ tag: "delim", value: ch, loc: { ...start, ...endLoc() } });
      continue;
    }

    // Unknown character
    return error(`unexpected character '${ch}'`, start);
  }

  tokens.push({ tag: "eof", value: "", loc: loc() });
  return tokens;
}

File: internal/lexer/lexer.go
// Package lexer implements the Clank lexer.
package lexer

import (
	"strings"

	"github.com/dalurness/clank/internal/token"
)

func isAlpha(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')
}

func isDigit(ch byte) bool {
	return ch >= '0' && ch <= '9'
}

func isIdentChar(ch byte) bool {
	return isAlpha(ch) || isDigit(ch) || ch == '_' || ch == '-'
}

func isWhitespace(ch byte) bool {
	return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r'
}

// Lex tokenizes source into a slice of tokens, or returns a LexError.
func Lex(source string) ([]token.Token, *token.LexError) {
	l := &lexer{source: source, line: 1, col: 1}
	return l.lex()
}

type lexer struct {
	source      string
	pos         int
	line        int
	col         int
	interpDepth int // >0 when inside ${...} within an interpolated string
}

func (l *lexer) loc() token.Loc {
	return token.Loc{Line: l.line, Col: l.col}
}

func (l *lexer) endLoc() (int, int) {
	return l.line, l.col
}

func (l *lexer) peek() byte {
	if l.pos < len(l.source) {
		return l.source[l.pos]
	}
	return 0
}

func (l *lexer) advance() byte {
	ch := l.source[l.pos]
	l.pos++
	if ch == '\n' {
		l.line++
		l.col = 1
	} else {
		l.col++
	}
	return ch
}

func (l *lexer) lexError(msg string, at *token.Loc) *token.LexError {
	loc := l.loc()
	if at != nil {
		loc = *at
	}
	start := l.pos - 20
	if start < 0 {
		start = 0
	}
	end := l.pos + 20
	if end > len(l.source) {
		end = len(l.source)
	}
	return &token.LexError{Code: "E001", Message: msg, Location: loc, Context: l.source[start:end]}
}

func (l *lexer) lex() ([]token.Token, *token.LexError) {
	var tokens []token.Token

	for l.pos < len(l.source) {
		ch := l.peek()

		// Whitespace
		if isWhitespace(ch) {
			l.advance()
			continue
		}

		// Comment
		if ch == '#' {
			for l.pos < len(l.source) && l.peek() != '\n' {
				l.advance()
			}
			continue
		}

		start := l.loc()

		// String literal (with interpolation support)
		if ch == '"' {
			l.advance() // opening quote
			toks, err := l.scanString(&start)
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, toks...)
			continue
		}

		// Number literal
		if isDigit(ch) {
			var num strings.Builder
			for l.pos < len(l.source) && isDigit(l.peek()) {
				num.WriteByte(l.advance())
			}
			if l.peek() == '.' && l.pos+1 < len(l.source) && isDigit(l.source[l.pos+1]) {
				num.WriteByte(l.advance()) // the dot
				for l.pos < len(l.source) && isDigit(l.peek()) {
					num.WriteByte(l.advance())
				}
				endLine, endCol := l.endLoc()
				tokens = append(tokens, token.Token{Tag: token.Rat, Value: num.String(), Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			} else {
				endLine, endCol := l.endLoc()
				tokens = append(tokens, token.Token{Tag: token.Int, Value: num.String(), Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			}
			continue
		}

		// Identifier or keyword
		if isAlpha(ch) || ch == '_' {
			var word strings.Builder
			for l.pos < len(l.source) && isIdentChar(l.peek()) {
				word.WriteByte(l.advance())
			}
			endLine, endCol := l.endLoc()
			w := word.String()
			loc := token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}
			if w == "true" || w == "false" {
				tokens = append(tokens, token.Token{Tag: token.Bool, Value: w, Loc: loc})
			} else if token.Keywords[w] {
				tokens = append(tokens, token.Token{Tag: token.Keyword, Value: w, Loc: loc})
			} else {
				tokens = append(tokens, token.Token{Tag: token.Ident, Value: w, Loc: loc})
			}
			continue
		}

		// Multi-char operators
		if l.pos+1 < len(l.source) {
			two := l.source[l.pos : l.pos+2]
			found := false
			for _, op := range token.MultiOps {
				if two == op {
					found = true
					break
				}
			}
			if found {
				l.advance()
				l.advance()
				endLine, endCol := l.endLoc()
				tokens = append(tokens, token.Token{Tag: token.Op, Value: two, Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
				continue
			}
		}

		// Range operators: ..= (3-char) and .. (2-char)
		if ch == '.' && l.pos+1 < len(l.source) && l.source[l.pos+1] == '.' {
			if l.pos+2 < len(l.source) && l.source[l.pos+2] == '=' {
				l.advance()
				l.advance()
				l.advance()
				endLine, endCol := l.endLoc()
				tokens = append(tokens, token.Token{Tag: token.Op, Value: "..=", Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			} else {
				l.advance()
				l.advance()
				endLine, endCol := l.endLoc()
				tokens = append(tokens, token.Token{Tag: token.Op, Value: "..", Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			}
			continue
		}

		// Single-char operators
		if token.SingleOps[ch] {
			l.advance()
			endLine, endCol := l.endLoc()
			tokens = append(tokens, token.Token{Tag: token.Op, Value: string(ch), Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			continue
		}

		// Delimiters
		if token.Delimiters[ch] {
			// When inside string interpolation, a } at depth 1 ends the interpolation
			if ch == '}' && l.interpDepth > 0 {
				l.advance()
				endLine, endCol := l.endLoc()
				tokens = append(tokens, token.Token{Tag: token.InterpEnd, Value: "}", Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
				l.interpDepth--
				// Resume scanning the rest of the string
				strStart := l.loc()
				toks, err := l.scanString(&strStart)
				if err != nil {
					return nil, err
				}
				tokens = append(tokens, toks...)
				continue
			}
			l.advance()
			endLine, endCol := l.endLoc()
			tokens = append(tokens, token.Token{Tag: token.Delim, Value: string(ch), Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			continue
		}

		// Unknown character
		return nil, l.lexError("unexpected character '"+string(ch)+"'", &start)
	}

	tokens = append(tokens, token.Token{Tag: token.EOF, Value: "", Loc: l.loc()})
	return tokens, nil
}

// scanString scans the body of a string literal (after the opening " or after
// an InterpEnd resumes). It returns tokens until the closing " or the next ${.
// On ${, it emits a Str token for the content so far, then an InterpStart
// token, increments interpDepth, and returns — the main loop lexes the
// expression, and when it hits the matching }, it calls scanString again.
func (l *lexer) scanString(start *token.Loc) ([]token.Token, *token.LexError) {
	var tokens []token.Token
	var buf strings.Builder

	for l.pos < len(l.source) && l.peek() != '"' {
		// Interpolation: ${
		if l.peek() == '$' && l.pos+1 < len(l.source) && l.source[l.pos+1] == '{' {
			// Emit accumulated string content
			endLine, endCol := l.endLoc()
			tokens = append(tokens, token.Token{Tag: token.Str, Value: buf.String(), Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
			buf.Reset()

			// Emit InterpStart and consume ${
			interpStart := l.loc()
			l.advance() // $
			l.advance() // {
			endLine, endCol = l.endLoc()
			tokens = append(tokens, token.Token{Tag: token.InterpStart, Value: "${", Loc: token.Loc{Line: interpStart.Line, Col: interpStart.Col, EndLine: endLine, EndCol: endCol}})
			l.interpDepth++
			return tokens, nil
		}

		// Escape sequences
		if l.peek() == '\\' {
			l.advance()
			if l.pos >= len(l.source) {
				return nil, l.lexError("unterminated string escape", start)
			}
			esc := l.advance()
			switch esc {
			case 'n':
				buf.WriteByte('\n')
			case 't':
				buf.WriteByte('\t')
			case '\\':
				buf.WriteByte('\\')
			case '"':
				buf.WriteByte('"')
			case '$':
				buf.WriteByte('$')
			default:
				return nil, l.lexError("invalid escape \\"+string(esc), start)
			}
			continue
		}

		buf.WriteByte(l.advance())
	}

	if l.pos >= len(l.source) {
		return nil, l.lexError("unterminated string", start)
	}

	l.advance() // closing "
	endLine, endCol := l.endLoc()
	tokens = append(tokens, token.Token{Tag: token.Str, Value: buf.String(), Loc: token.Loc{Line: start.Line, Col: start.Col, EndLine: endLine, EndCol: endCol}})
	return tokens, nil
}
